extensions/amp-story/1.0/amp-story-tooltip.js

```diff
@@ -0,0 +1,401 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  Action,
+  StateProperty,
+  UIType,
+  getStoreService,
+} from './amp-story-store-service';
+import {CSS} from '../../../build/amp-story-tooltip-1.0.css';
+import {EventType, dispatch} from './events';
+import {Services} from '../../../src/services';
+import {
+  addAttributesToElement,
+  childElement,
+  removeElement,
+} from '../../../src/dom';
+import {createShadowRootWithStyle, getSourceOriginForElement} from './utils';
+import {dev} from '../../../src/log';
+import {dict} from '../../../src/utils/object';
+import {getAmpdoc} from '../../../src/service';
+import {htmlFor, htmlRefs} from '../../../src/static-template';
+import {parseUrlDeprecated} from '../../../src/url';
+import {setImportantStyles} from '../../../src/style';
+
+/**
+ * Minimum vertical space needed to position tooltip.
+ */
+const MIN_VERTICAL_SPACE = 56;
+
+/**
+ * Padding between tooltip and edges of screen.
+ */
+const EDGE_PADDING = 8;
+
+/**
+ * List of selectors that can trigger a tooltip.
+ * @const {!Array<string>}
+ */
+export const TOOLTIP_TRIGGERABLE_SELECTORS = ['a[href]'];
+
+/**
+ * @struct @typedef {{
+ *   tooltip: !Element,
+ *   buttonLeft: !Element,
+ *   buttonRight: !Element,
+ *   arrow: !Element,
+ * }}
+ */
+let tooltipElementsDef;
+
+/**
+ * Tooltip element triggered by clickable elements in the amp-story-grid-layer.
+ */
+export class AmpStoryTooltip {
+  /**
+   * @param {!Window} win
+   * @param {!Element} storyEl
+   */
+  constructor(win, storyEl) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {!Element} */
+    this.storyEl_ = storyEl;
+
+    /** @private {boolean} */
+    this.isBuilt_ = false;
+
+    /** @private {?Element} */
+    this.shadowRoot_ = null;
+
+    /** @private {?Element} */
+    this.tooltipOverlayEl_ = null;
+
+    /** @private {?Element} */
+    this.tooltip_ = null;
+
+    /** @private {?Element} */
+    this.tooltipArrow_ = null;
+
+    /** @private @const {!./amp-story-store-service.AmpStoryStoreService} */
+    this.storeService_ = getStoreService(this.win_);
+
+    /** @private @const {!../../../src/service/resources-impl.Resources} */
+    this.resources_ = Services.resourcesForDoc(getAmpdoc(this.win_.document));
+
+    /**
+     * Clicked element producing the tooltip. Used to avoid building the same
+     * element twice.
+     * @private {?Element} */
+    this.clickedElem_ = null;
+
+    this.storeService_.subscribe(StateProperty.TOOLTIP_ELEMENT, elem => {
```

Comment:
Nit: please use either `element` or `el`. I don't mind `elem` but you're mixing both in this file (`storyEl`, `clickedElem`), and I think we mostly use `el` in the rest of the codebase.